src/utils/ucc_coll_utils.c

```diff
@@ -94,6 +94,68 @@ ucc_coll_args_is_mem_symmetric(const ucc_coll_args_t *args,
     return 0;
 }
 
+
+/* If this is the root and the src/dst buffers are asymmetric, the dst needs
+   to have a new allocation to make the mem types match. On task completion,
+   copy the result back into the old dst */
+ucc_status_t
+ucc_coll_args_init_asymmetric_buffer(ucc_coll_args_t *args,
+                                     ucc_team_h team,
+                                     ucc_buffer_info_asymmetric_memtype_t *save_info)
+{
+    ucc_status_t status = UCC_OK;
+
+    if (UCC_IS_INPLACE(*args)) {
+        return UCC_ERR_INVALID_PARAM;
+    }
+    switch (args->coll_type) {
+    case UCC_COLL_TYPE_REDUCE:
+    case UCC_COLL_TYPE_GATHER:
+    case UCC_COLL_TYPE_SCATTER:
```

Comment:
why can we combine gather and scatter here? if memory type is asymmetric then
for scatter:
src buffer valid only for root, we should use dst buffer memory type and replace src buffer with scratch 
for gather:
dst buffer valid only for root, we should use src buffer memory type and replace dst buffer with scratch